Update Function for Production Deployment

mcdonnnj opened the pull request:
This PR updates the lambda for deployment to the production environment. Changes include:
* Use hard coded directory names and changed `success_folder` and `error_folder` to `save_succeeded` and `save_failed` respectively. These values now represent booleans to determine if the input should be saved.
* Remove checking object name in lambda as it will be handled by the lambda's trigger instead.
* Reduce document validation down to the existence of two required fields instead of checking all fields against a list of allowed fields.
* Streamline database insertion into one method call.
* Track the number of documents actually processed.

mcdonnnj committed:
Extensive reworking:
-Corrected the collection being accessed on the database.
-Changed to loading field mapping JSON directly instead of copying the source file locally.
-Changed valid field source to JSON loaded directly from S3 instead of database querying.
-success_folder turned into boolean save_succeeded to mark whether or not to save with the directory set in the lambda code.
-error_folder changed the same way as success_folder with the new parameter called save_failed.
-No longer do extensive file validation in lambda handler. The file_suffix environment value should be the same value used to set a suffix match in the lambda configuration on AWS. Since another branch will change how the RVA ID is determined this simplifies file validation in advance. This condensed three parameters down into one that is only used in the handler.

mcdonnnj committed:
Update README.md to reflect some flag changes and clean up some f-string changed from blackened source in the lambda.

mcdonnnj committed:
Tweak handler to convert Terraform booleans into Python booleans.
Fix reading from S3 object to correctly import JSONs.

mcdonnnj committed:
Update to no longer check valid fields. The presence of the "RVA ID" and "NCATS ID" fields signifies a finding to add to the collection.

mcdonnnj committed:
Remove valid_fields reference in lambda handler.

mcdonnnj committed:
Adjust where rvaId reassignment occurs. Now track successfully processed documents.

mcdonnnj committed:
Consolidate into one database method call using the upsert option for the find_one_and_update() method.

mcdonnnj committed:
Removed references to valid fields due to changed logic.

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -19,15 +19,13 @@
+                              the database.
+  --save-failed=FAILED        The directory name used for storing unsuccessfully
+                              processed files. [default: True]
+  --save-succesded=SUCCEEDED  The directory name used for storing successfully
```
Typo in `--save-succeeded`

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -153,21 +144,20 @@ def import_data(
-        s3_client.download_file(
-            Bucket=s3_bucket, Key=fields_filename, Filename=temp_field_data_filepath
+        # Fetch object for the field_map JSON
+        field_map_object = s3_client.get_object(Bucket=s3_bucket, Key=field_map)
```
It may be worth logging a warning if the `field_map_object` is `None`.

mcdonnnj replied:
Tested and if the object does not exist an Access Denied error is thrown in the `get_object()` call.

dav3r replied:
👍

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -194,97 +184,91 @@ def import_data(
+
+            # Only process appropriate findings records.
+            if "RVA ID" in finding.keys() and "NCATS ID" in finding.keys():
+                finding["RVA ID"] = rvaId
```
Do we need to do any validation on `rvaId` to attempt to ensure it's correct or do we take care of that elsewhere?

mcdonnnj replied:
This is assumed to be taken care of manually in the filename. @mzack5020 has a branch he is working on to derive this from the JSON data on a per finding basis.

dav3r replied:
This is probably worth mentioning in a comment earlier where `rvaId` is initially set.

jsf9k replied:
I agree with @dav3r here.

mcdonnnj replied:
Resolved in 320a0bc ,  thanks for the feedback.

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -194,97 +184,91 @@ def import_data(
+        if save_succeeded:
+            # Create success folders depending on how processing went
+            succeeded_filename = data_filename.replace(
+                ".json", f"_{str(datetime.datetime.now())}.json"
```
I'm not a huge fan of filenames (or S3 keys in this case) with spaces in them.  Perhaps use `strftime` to format to the timestamp?

Same thing below in the `save_failed` section.

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -194,97 +184,91 @@ def import_data(
-        )
+            logging.info(
+                f"Moved {data_filename} to the success directory under folder "
+                f"name {SUCCEEDED_FOLDER}"
```
This logging message should also mention `{succeeded_filename}` in some way (e.g. `"Moved {data_filename} to the success directory under folder name {SUCCEEDED_FOLDER} as {succeeded_filename}"`).

Same thing below in the `save_failed` section.

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -194,97 +184,91 @@ def import_data(
+                f"name {SUCCEEDED_FOLDER}"
+            )
+    except Exception as err:
+        logging.warning(f"Error Message {type(err)}: {err}")
```
Why not `logging.error` here?

dav3r commented on fdi/findings_data_import.py:
```diff
@@ -194,97 +184,91 @@ def import_data(
+            try:
+                s3_client.delete_object(Bucket=s3_bucket, Key=data_filename)
+            except ClientError as delete_error:
+                logging.info(f"Error deleting file with error: {delete_error}")
```
Why not `logging.error` here?

dav3r reviewed (changes requested):
No major issues, but I listed some smaller things that should be updated or at least considered.

mcdonnnj committed:
Updates per feedback:
Added a debug message to output the field_map dictionary after import.
Fixed incorrect logging classifications.
Changed timestamp to output per strftime() formatting to eliminate space.
Added object name in output for when a file succeeds or fails.

mzack5020 commented:
We went over this in person on Wednesday so it was an easy review. Thank you, Nick.

jsf9k reviewed (approved):
This looks good to me.  I do agree with @dav3r about the need to add a comment in one place.

mcdonnnj committed:
Add comment explaining filename format for RVA ID extraction per feedback.
